CreeDictionary/utils/crkeng_xml_utils.py

```diff
@@ -79,3 +106,237 @@ def convert_xml_inflectional_category_to_word_class(
+        }
+
+        source_abbreviations = []
+
+        for s in root.findall(".//source"):
+            abbreviation = s.get("id")
+            assert abbreviation is not None
+            source_abbreviations.append(abbreviation)
+
+        return cls(entries=entries, source_abbreviations=source_abbreviations)
+
+    @classmethod
+    def _build_indexes(cls, entries: Set[XMLEntry]) -> Indexes:
+        """
+        build in-memory indexes for fast querying later
+        """
+        # first classify the entries by each field names
+        # then we merge them by the key combinations needed
+
+        # helper type hint
+        _SingleKeyIndexes = Dict[
+            NamedTupleFieldName, Dict[NamedTupleFieldValue, Set[XMLEntry]]
+        ]
+
+        single_key_indexes: _SingleKeyIndexes = {}
+        for field_name in set(chain(*cls.INDEX_KEYS)):
+            field_value_to_entry = defaultdict(set)
+            for entry in entries:
+                field_value_to_entry[getattr(entry, field_name)].add(entry)
+            single_key_indexes[field_name] = field_value_to_entry
+
+        indexes: Dict[
+            Tuple[NamedTupleFieldName, ...],
+            Dict[Tuple[NamedTupleFieldValue, ...], Set[XMLEntry]],
+        ] = {tuple(sorted(field_names)): {} for field_names in cls.INDEX_KEYS}
+
+        for sorted_field_names, values_to_entries in indexes.items():
+
+            for entry in entries:
+                values_to_entries[
+                    tuple(
+                        getattr(entry, field_name) for field_name in sorted_field_names
+                    )
+                ] = set.union(
+                    *(
+                        single_key_indexes[field_name][getattr(entry, field_name)]
+                        for field_name in sorted_field_names
+                    )
+                )
+
+        return indexes
+
+    def __init__(self, entries: Set[XMLEntry], source_abbreviations: Iterable[str]):
+        self._entries = entries
+        self._source_abbreviations = source_abbreviations
+
+        self._indexes = self._build_indexes(entries)
+
+    def filter(self, **constraints: NamedTupleFieldValue) -> Set[XMLEntry]:
+        """
+        A lookup method that mimics django's queryset API. Except the relevant indexes must be built in advance.
+        The returned set is empty when no matching entries are found
+
+        :raise ValueError: when the relevant index does not exist for the query
+        """
+        # sort by field names
+        sorted_constraints = sorted(constraints.items(), key=lambda x: x[0])
+        field_names, field_values = zip(*sorted_constraints)
+
+        try:
+            values_to_entries = self._indexes[
+                cast(Tuple[NamedTupleFieldName], field_names,)
+            ]
+        except ValueError:
+            raise ValueError(f"index with field names {list(constraints)} is not built")
+
+        return values_to_entries.get(field_values, set())
```

Comment:
As commented above, it turns out `filter(l="xxx")` is ever needed.

Reply:
Do note and be prepared for that the `<stem>...</stem>` field is empty for some two thousand lexical/dictionary entries, because they are phrases or the CW source provides no such stem.